File: net/OpenDraft.Lang.UnitTest/TokenReaderTestUtility.cs
// <copyright file="TokenReaderTestUtility.cs" company="Jon Rowlett">
// Copyright (c) Jon Rowlett. All rights reserved.
// </copyright>

namespace OpenDraft.Lang.UnitTest;

using System.Runtime.CompilerServices;

/// <summary>
/// Utility methods for testing the <see cref="TokenReader"/> class.
/// </summary>
internal static class TokenReaderTestUtility
{
    /// <summary>
    /// Creates a <see cref="TokenReader"/> for the given input string, using the given log and source reference information.
    /// </summary>
    /// <param name="input">The input string to read tokens from.</param>
    /// <param name="log">The log to record messages to.</param>
    /// <param name="filePath">The source file path for the input string.</param>
    /// <param name="lineNumber">The line number in the source file for the input string.</param>
    /// <returns>A <see cref="TokenReader"/> instance for the given input string.</returns>
    public static TokenReader CreateTokenReader(
        string input,
        TestMessageLog log,
        [CallerFilePath] string filePath = "",
        [CallerLineNumber] int lineNumber = 0)
    {
        return new TokenReader(
            new StringReader(input),
            new SourceReference(filePath, lineNumber),
            log.Log);
    }

    /// <summary>
    /// Creates a <see cref="TokenReader"/> for the given input string, using a new log and source reference information.
    /// </summary>
    /// <param name="expectedValue">The expected value of the single token.</param>
    /// <param name="input">The input string to read tokens from.</param>
    /// <remarks>
    /// This method asserts that the input string produces a single token with the expected value and that the token reader is clean afterwards.
    /// </remarks>
    public static void SingleTokenCleanTest(
        string expectedValue,
        string input)
    {
        var log = new TestMessageLog();

        TokenReader reader = CreateTokenReader(input, log);
        var token = reader.Read();
        Assert.IsNotNull(token);
        Assert.IsInstanceOfType<StringLiteralToken>(token);
        Assert.AreEqual(expectedValue, ((StringLiteralToken)token).Value);

        Assert.IsNull(reader.Read());
        log.AssertClean();
    }

    /// <summary>
    /// Creates a <see cref="TokenReader"/> for the given input string, using a new log and source reference information.
    /// </summary>
    /// <param name="expectedValue">The expected value of the single token.</param>
    /// <param name="input">The input string to read tokens from.</param>
    /// <param name="unit">The expected unit of the numeric literal, if any.</param>
    /// <remarks>
    /// This method asserts that the input string produces a single token with the expected value and that the token reader is clean afterwards.
    /// </remarks>
    public static void SingleTokenCleanTest(
        long expectedValue,
        string input,
        string? unit = null)
    {
        var log = new TestMessageLog();

        TokenReader reader = CreateTokenReader(input, log);
        var token = reader.Read();
        Assert.IsNotNull(token);
        Assert.IsInstanceOfType<NumericLiteralToken>(token);
        var numToken = (NumericLiteralToken)token;
        Assert.IsFalse(numToken.IsFloatingPoint);
        Assert.AreEqual(expectedValue, numToken.IntegerValue);
        Assert.AreEqual(unit, numToken.Unit);

        Assert.IsNull(reader.Read());
        log.AssertClean();
    }

    /// <summary>
    /// Creates a <see cref="TokenReader"/> for the given input string, using a new log and source reference information.
    /// </summary>
    /// <param name="expectedValue">The expected value of the single token.</param>
    /// <param name="input">The input string to read tokens from.</param>
    /// <param name="unit">The expected unit of the numeric literal, if any.</param>
    /// <remarks>
    /// This method asserts that the input string produces a single token with the expected value and that the token reader is clean afterwards.
    /// </remarks>
    public static void SingleTokenCleanTest(
        decimal expectedValue,
        string input,
        string? unit = null)
    {
        var log = new TestMessageLog();

        TokenReader reader = CreateTokenReader(input, log);
        var token = reader.Read();
        Assert.IsNotNull(token, input);
        Assert.IsInstanceOfType<NumericLiteralToken>(token, input);
        var numToken = (NumericLiteralToken)token;
        Assert.IsTrue(numToken.IsFloatingPoint, input);
        Assert.AreEqual(expectedValue, numToken.FloatingPointValue);
        Assert.AreEqual(unit, numToken.Unit);

        Assert.IsNull(reader.Read(), input);
        log.AssertClean();
    }
}

File: cpp/test/lang_ut/token_reader_test_utility.cpp
/**
 * Copyright (c) Jon Rowlett. All rights reserved.
 * Licensed under the MIT license. See LICENSE file in the project root for full license information.
 */

#include "token_reader_test_utility.hpp"

using namespace opendraft::lang;
using namespace std;
using namespace testfx;

shared_ptr<token_reader> token_reader_test_utility::create_token_reader(
    std::istream& input,
    test_message_log& log,
    const std::source_location& location)
{
    source_reference start = source_reference::from_source_location(location);
    return make_shared<token_reader>(
        input,
        start,
        [&log](const message& msg) { log.log(msg); });
}

void token_reader_test_utility::single_token_clean_test(
    const std::string& expected_value,
    const std::string& input,
    const std::source_location& location)
{
    test_message_log log;
    std::istringstream ss(input);
    auto reader = create_token_reader(ss, log, location);
    auto token = reader->read();
    assert_not_null(token, location);
    auto type = token->type();
    assert_equal(token::token_type::STRING_LITERAL, type, location);
    auto string_token = static_cast<string_literal_token*>(token.get());
    assert_equal(string(expected_value), string_token->value(), location);
    assert_equal((size_t)0, log.messages().size(), location);
}

void token_reader_test_utility::single_token_clean_test(
    long expected_value,
    const std::string& input,
    const char* expected_units,
    const std::source_location& location)
{
    test_message_log log;
    std::istringstream ss(input);
    auto reader = create_token_reader(ss, log, location);
    auto token = reader->read();
    assert_not_null(token, location);
    auto type = token->type();
    assert_equal(token::token_type::NUMERIC_LITERAL, type, location);
    auto numeric_token = static_cast<numeric_literal_token*>(token.get());
    assert_equal(true, numeric_token->is_integer());
    assert_equal(expected_value, numeric_token->integer_value(), location);

    if (expected_units != nullptr)
    {
        assert_equal(string(expected_units), numeric_token->unit(), location);
    }

    assert_equal((size_t)0, log.messages().size(), location);
}

void token_reader_test_utility::single_token_clean_test(
    double expected_value,
    const std::string& input,
    const char* expected_units,
    const std::source_location& location)
{
    test_message_log log;
    std::istringstream ss(input);
    auto reader = create_token_reader(ss, log, location);
    auto token = reader->read();
    assert_not_null(token, location);
    auto type = token->type();
    assert_equal(token::token_type::NUMERIC_LITERAL, type, location);
    auto numeric_token = static_cast<numeric_literal_token*>(token.get());
    assert_equal(false, numeric_token->is_integer(), location);
    assert_equal(expected_value, numeric_token->double_value(), location);

    if (expected_units != nullptr)
    {
        assert_equal(string(expected_units), numeric_token->unit(), location);
    }

    assert_equal((size_t)0, log.messages().size(), location);
}